mlir/test/Integration/Dialect/Tosa/CPU/test-maxpool-dynamic.mlir

```diff
@@ -0,0 +1,87 @@
+// RUN:   mlir-opt %s \
+// RUN:      -pass-pipeline="builtin.module(func.func(tosa-infer-shapes,tosa-to-linalg-named,tosa-to-linalg,tosa-to-arith))" \
+// RUN: | mlir-opt \
+// RUN:      -one-shot-bufferize="bufferize-function-boundaries" \
+// RUN:      -buffer-deallocation-pipeline -test-lower-to-llvm \
+// RUN: | mlir-cpu-runner \
+// RUN:     -entry-point-result=void \
+// RUN:     -shared-libs=%mlir_c_runner_utils \
+// RUN: | FileCheck %s
+
+// Validate that the TOSA lowering for tosa.max_pool2d produces the same results when
+// for fully static and fully dynamic inputs.
+
+!Generator = !llvm.ptr
+
+// Utility functions
+llvm.func @printI1(i1)
+func.func private @rtsrand(index) -> (!Generator)
+func.func private @rtrand(!Generator, i64) -> (i64)
+func.func private @rtdrand(!Generator) -> ()
+
+func.func @max_pool_static(%arg0: tensor<1x6x34x62xf32>) -> (tensor<1x6x34x62xf32>) {
+  %0 = tosa.max_pool2d %arg0 {
+    pad = array<i64: 1, 1, 1, 1>,
+    kernel = array<i64: 3, 3>,
+    stride = array<i64: 1, 1>
+  } : (tensor<1x6x34x62xf32>) -> tensor<1x6x34x62xf32>
+  return %0 : tensor<1x6x34x62xf32>
+}
+
+func.func @max_pool_dynamic(%arg0: tensor<?x?x?x?xf32>) -> (tensor<?x?x?x?xf32>) {
+  %0 = tosa.max_pool2d %arg0 {
+    pad = array<i64: 1, 1, 1, 1>,
+    kernel = array<i64: 3, 3>,
+    stride = array<i64: 1, 1>
+  } : (tensor<?x?x?x?xf32>) -> tensor<?x?x?x?xf32>
+  return %0 : tensor<?x?x?x?xf32>
+}
+
+// Test harness to compare the results of a fully statically shaped max_pool2d with
+// a fully dynamically shaped max_pool2d on the same inputs.
+func.func @main() {
```

Comment:
Rather than a test that compares the behavior of static and dynamic dimensions for randomly generated inputs, I'd find more value in a test comparing against a reference output for deterministic inputs to catch possible bugs affecting both static and dynamic paths, similar to `test-fully-connected.mlir`. Maybe one test for static + one test for dynamic.

Reply:
That was a good idea, it ended up catching a bug in the implementation.